harness/determined/pytorch/samplers.py

```diff
@@ -0,0 +1,361 @@
+"""
+Guidelines for Reproducible Datasets.
+
+.. note::
+
+   If you are using a MapDataset and not an IterableDataset, make your life easy and use the
+   context.TrainingDataLoader() and context.ValidationDataLoader(), which are a drop-in replacement
+   for pytorch's built-in DataLoader, and Determined will handle all of these details.
+
+Achieving a reproducible dataset that is able to pause and continue (sometimes called "incremental
+training") is easy if you follow a few rules.
+
+- Even if you are going to ultimate return an IterableDataset, it is best to use PyTorch's Sampler
+  class as the basis for choosing the order of which records to read.  Operations on Samplers are
+  quick and cheap, while operations on data afterwards are expensive.  For more details, see the
+  discussion at TODO:ref_yogadl_readme.  If you don't have a custom sampler, start with a simple
+  one:
+
+  .. code::python
+
+     sampler = torch.utils.data.SequentialSampler(my_dataset)
+
+- **Shuffle first**: Always use a reproducible shuffle when you shuffle. Determined provides a
+  ReproducibleShuffleSampler for this purpose.  Determined also provides a
+  ReproducibleShuffleBatchSampler class, but you should prefer to shuffle at the record level
+  whenever possible, to achieve the highest-quality shuffle.
```

Comment:
"but you should prefer to shuffle at the record level" - wait shouldnt the determined shuffler do that? Would it be better to say "if you write a custom sampler, you should shuffle at the record level ...."

Reply:
This is talking about the difference between the ReproducibleShuffleSampler and the ReproducibleShuffleBatchSampler.  I'll try to make that more obvious in the text.